src/gar.lua

```diff
@@ -1577,4 +1577,271 @@ function gar.getPaginatedDelegations(address, cursor, limit, sortBy, sortOrder)
+				"Quantity must be less than or equal to the vaulted stake amount."
+			)
+
+			if existingVault.balance == stakeToTakeFromSource then
+				-- The operator vault has been emptied
+				sourceGateway.vaults[vaultId] = nil
+			else
+				-- The operator vault has been partially emptied
+				sourceGateway.vaults[delegateAddress][vaultId].balance = sourceGateway.vaults[delegateAddress][vaultId].balance
+					- stakeToTakeFromSource
+			end
+		else
+			-- Get the redelegation amount from the operator stakes
+			local maxWithdraw = sourceGateway.operatorStake - gar.getSettings().operators.minStake
+			assert(
+				stakeToTakeFromSource <= maxWithdraw,
+				"Resulting stake is not enough to maintain the minimum operator stake of "
+					.. gar.getSettings().operators.minStake
+					.. " IO"
+			)
+
+			sourceGateway.operatorStake = sourceGateway.operatorStake - stakeToTakeFromSource
+		end
+	else
+		local existingDelegate = sourceGateway.delegates[delegateAddress]
+		assert(existingDelegate, "This delegate has no stake to redelegate.")
+
+		if vaultId then
+			local existingVault = existingDelegate.vaults[vaultId]
+			assert(existingVault, "Vault not found on the delegate.")
+			assert(
+				existingVault.balance >= stakeToTakeFromSource,
+				"Quantity must be less than or equal to the vaulted stake amount."
+			)
+
+			if existingVault.balance == stakeToTakeFromSource then
+				-- The vault has been emptied
+				sourceGateway.delegates[delegateAddress].vaults[vaultId] = nil
+
+				if
+					-- If all vaults and stake are emptied, remove the delegate
+					sourceGateway.delegates[delegateAddress].delegatedStake == 0
+					and next(sourceGateway.delegates[delegateAddress].vaults) == nil
+				then
+					gar.pruneDelegateFromGateway(delegateAddress, sourceGateway)
+				end
+			else
+				-- The vault has been partially emptied
+				sourceGateway.delegates[delegateAddress].vaults[vaultId].balance = sourceGateway.delegates[delegateAddress].vaults[vaultId].balance
+					- stakeToTakeFromSource
+			end
+		else
+			-- Check if the delegate has enough stake to redelegate
+			assert(
+				existingDelegate.delegatedStake >= stakeToTakeFromSource,
+				"Quantity must be less than or equal to the delegated stake amount."
+			)
+
+			-- Check if the delegate will have enough stake left after re-delegating
+			local existingStake = existingDelegate.delegatedStake
+			local requiredMinimumStake = sourceGateway.settings.minDelegatedStake
+			local maxAllowedToWithdraw = existingStake - requiredMinimumStake
+			assert(
+				stakeToTakeFromSource <= maxAllowedToWithdraw or stakeToTakeFromSource == existingStake,
+				"Remaining delegated stake must be greater than the minimum delegated stake. Adjust the amount or re-delegate all stake."
+			)
+
+			-- If the delegate has enough stake to redelegate, move the stake. If its all the stake, remove the delegate
+			if existingDelegate.delegatedStake == stakeToTakeFromSource then
```

Comment:
This is only appropriate if the delegate has no withdraw vaults. We'll have to fix this.

Reply:
Fixed